### PR TITLE
Migrates to langgraph

### PR DESCRIPTION
**Pull request by filipeximenes**
This PR sunsets the previous use of chains in favor of the more modern, more readable and simpler to use langgraph approach. The migration was done trying to keep the changes as minimal as possible and the both methods `as_chain` and `as_graph` are still available. For now `as_chain` still the default approach but it can be changed by changing the setting `AI_ASSISTANT_USE_LANGGRAPH` to `True`. After this changes have been field tested we will sunset the old `as_chain` approach. Tests are already running with `AI_ASSISTANT_USE_LANGGRAPH` set to true. During this migration only 5 tests broke because it was required rerunning the cassettes due to small changes in the setup of how messages are organized in the prompt.

### CODE REVIEW

File: django_ai_assistant/helpers/assistants.py [hunk 1]
```diff
@@ -577,7 +701,7 @@ def as_tool(self, description: str) -> BaseTool:
         Returns:
             BaseTool: A tool that runs the assistant. The tool name is this assistant's id.
         """
-        return Tool.from_function(
+        return StructuredTool.from_function(
```
Comment (fjsj) [inline django_ai_assistant/helpers/assistants.py:704]: Is this change necessary?

File: django_ai_assistant/langchain/chat_message_histories.py [hunk 2]
```diff
@@ -73,13 +77,17 @@ def add_messages(self, messages: Sequence[BaseMessage]) -> None:
             messages: A list of BaseMessage objects to store.
         """
         with transaction.atomic():
+            existing_messages = self.get_messages()
+
+            messages_to_create = [m for m in messages if m not in existing_messages]
```
Comment (fjsj) [inline django_ai_assistant/langchain/chat_message_histories.py:82]: how is the `not in` making the equality for messages? do you know?

File: django_ai_assistant/helpers/assistants.py [hunk 3]
```diff
@@ -437,6 +449,115 @@ def get_history_aware_retriever(self) -> Runnable[dict, RetrieverOutput]:
...
+            output: str
+
+        def setup(state: AgentState):
+            return {"messages": [SystemMessage(content=self.get_instructions())]}
+
+        def retriever(state: AgentState):
+            if not self.has_rag:
+                return
+
+            retriever = self.get_history_aware_retriever()
```
Comment (fjsj) [inline django_ai_assistant/helpers/assistants.py:499]: Seems the `get_history_aware_retriever` is still using `RunnableBranch`, to summarize the previous conversation when has_rag = True and the user is making a second question. That's fine, but we may want to change this to use langgraph in the next PR. LMK if you need clarification on this.

File: example/tour_guide/ai_assistants.py [hunk 4]
```diff
@@ -34,22 +34,17 @@ class TourGuideAIAssistant(AIAssistant):
...
+        "Your response should only contain valid JSON data. DON'T include '```json' in your response. "
         "The JSON should be formatted according to the following structure: \n"
         f"\n\n{_tour_guide_example_json()}\n\n\n"
         "In the 'attraction_name' field provide the name of the attraction in english. "
         "In the 'attraction_description' field generate an overview about the attraction with the most important information, "
         "curiosities and interesting facts. "
         "Only include a value for the 'attraction_url' field if you find a real value in the provided data otherwise keep it empty. "
     )
-    model = "gpt-4o"
+    model = "gpt-4o-2024-08-06"
```
Comment (fjsj) [inline example/tour_guide/ai_assistants.py:47]: Perhaps its best to not hardcode the date here?

File: tests/test_helpers/test_assistants.py [hunk 5]
```diff
@@ -92,24 +92,24 @@ def test_AIAssistant_invoke():
...
+    )
+
+    question_message = response_1["messages"][1]
+    assert question_message.content == "What is the temperature today in Recife?"
+    assert question_message.id == str(messages_ids[0])
+    response_message = response_1["messages"][2]
+    assert (
+        response_message.content == "The current temperature in Recife today is 32 degrees Celsius."
+    )
+    assert response_message.id == str(messages_ids[1])
```
Comment (fjsj) [inline tests/test_helpers/test_assistants.py:110]: I preferred the old approach for readability. If you had trouble with the diffs, it's possible to change pytest to include all the diff.

File: tests/test_views.py [hunk 6]
```diff
@@ -301,7 +301,7 @@ def test_create_thread_message(authenticated_client):
     )
     assert (
         ai_message.message["data"]["content"]
-        == "The current temperature in San Francisco, CA is 32 degrees Celsius."
+        == "The current temperature in San Francisco, CA is 32°C."
```
Comment (fjsj) [inline tests/test_views.py:304]: If this changed, perhaps you need to also update the vcr recordings when langgraph is false?

**Review by fjsj — CHANGES_REQUESTED** [review]
Overall LGTM, but please check comments.
Can you also check the coverage decrease?

File: tests/test_views.py (hunk 6 above)
Comment (fjsj) [reply tests/test_views.py:304]: Ah, I think it won't be consistent on two runs... so the test can only run with true or false. Another reason to remove the old code.

File: django_ai_assistant/helpers/assistants.py (hunk 1 above)
Comment (filipeximenes) [reply django_ai_assistant/helpers/assistants.py:704]: Yes, that's due to the update to `langchain-core==0.2.38` required to use `langgraph`. Running with `Tool` gives an error due to tool parameters not aligning and the error itself recommends using `StructuredTool`

File: django_ai_assistant/langchain/chat_message_histories.py (hunk 2 above)
Comment (filipeximenes) [reply django_ai_assistant/langchain/chat_message_histories.py:82]: Good point, it worked but I didn't check why it was working, perhaps we should force comparing by id to make this safer and more explicit?

File: django_ai_assistant/helpers/assistants.py (hunk 3 above)
Comment (filipeximenes) [reply django_ai_assistant/helpers/assistants.py:499]: Yes, this was intentional to keep things simpler for now, definitely something we should migrate as well later

File: example/tour_guide/ai_assistants.py (hunk 4 above)
Comment (filipeximenes) [reply example/tour_guide/ai_assistants.py:47]: In this case it makes sense as structured outputs was trained in this version

1 https://openai.com/index/introducing-structured-outputs-in-the-api/
```
A new option for the response_format parameter: developers can now supply a JSON Schema via json_schema, a new option for the response_format parameter. This is useful when the model is not calling a tool, but rather, responding to the user in a structured way. This feature works with our newest GPT-4o models: gpt-4o-2024-08-06, released today, and gpt-4o-mini-2024-07-18. When a response_format is supplied with strict: true, model outputs will match the supplied schema.
```

https://platform.openai.com/docs/models/gpt-4o
```
Latest snapshot that supports [Structured Outputs](https://platform.openai.com/docs/guides/structured-outputs)
```

File: example/tour_guide/ai_assistants.py [hunk 7]
```diff
@@ -34,22 +34,17 @@ class TourGuideAIAssistant(AIAssistant):
...
-        "Use the available tools to suggest nearby attractions to the user. "
-        "You don't need to include all the found items, only include attractions that are relevant for a tourist. "
-        "Select the top 10 best attractions for a tourist, if there are less then 10 relevant items only return these. "
-        "Order items by the most relevant to the least relevant. "
-        "If there are no relevant attractions nearby, just keep the list empty. "
-        "Your response will be integrated with a frontend web application therefore it's critical that "
-        "it only contains a valid JSON. DON'T include '```json' in your response. "
+        "You will receive the user coordinates and should use available tools to find nearby attractions. "
+        "Only call the find_nearby_attractions tool once. "
+        "Your response should only contain valid JSON data. DON'T include '```json' in your response. "
```
[the remark this answers is not in the document]
Comment (filipeximenes) [reply example/tour_guide/ai_assistants.py:39]: I was having issues where gpt would enter a loop calling the same tool, this new prompt seems to be working better

File: tests/test_helpers/test_assistants.py (hunk 5 above)
Comment (filipeximenes) [reply tests/test_helpers/test_assistants.py:110]: Yes, it was better, the problem is that in the new version of langchain-core there's a bunch of metadata (see example bellow) that is not relevant to tests added to messages:
Screenshot 2024-09-11 at 14 11 45

File: tests/test_views.py (hunk 6 above)
Comment (filipeximenes) [reply tests/test_views.py:304]: Yep, that's it, the vcr had to be updated so there were small changes to outputs

File: django_ai_assistant/helpers/assistants.py (hunk 3 above)
Comment (filipeximenes) [reply django_ai_assistant/helpers/assistants.py:499]: Created an issue for it https://github.com/vintasoftware/django-ai-assistant/issues/166

File: django_ai_assistant/langchain/chat_message_histories.py (hunk 2 above)
Comment (filipeximenes) [reply django_ai_assistant/langchain/chat_message_histories.py:82]: The reason why this works is mainly because the built in `add_messages` function that is called before this method is called does the heavy lifting of returning just the messages that are new. So the code here is actually wrong, I'll return it to the previous state so there's no risk of someone thinking this is doing more than it actually is.

Comment (fjsj) [reply django_ai_assistant/langchain/chat_message_histories.py:82]: I think so, but does the messages Langchain passes as params contain any ids?
Also, we can optimize this a bit by querying with ids.

File: django_ai_assistant/helpers/assistants.py (hunk 3 above)
Comment (fjsj) [reply django_ai_assistant/helpers/assistants.py:499]: Thanks!

File: example/tour_guide/ai_assistants.py (hunk 4 above)
Comment (fjsj) [reply example/tour_guide/ai_assistants.py:47]: sounds good

File: django_ai_assistant/langchain/chat_message_histories.py [hunk 8]
```diff
@@ -73,13 +77,17 @@ def add_messages(self, messages: Sequence[BaseMessage]) -> None:
             messages: A list of BaseMessage objects to store.
         """
         with transaction.atomic():
+            existing_message_ids = [m.id for m in self.get_messages()]
+
+            messages_to_create = [m for m in messages if m.id not in existing_message_ids]
```
Comment (filipeximenes) [inline django_ai_assistant/langchain/chat_message_histories.py:82]: @fjsj although this was working I've update it to compare by ID, I think this makes it safer to internal changes

File: tests/test_helpers/test_assistants.py (hunk 5 above)
Comment (fjsj) [reply tests/test_helpers/test_assistants.py:110]: Got it! We could use something like this: https://kalnytskyi.com/posts/assert-str-matches-regex-in-pytest/
In this case, it would be a comparator that always returns true.
But that's a nitpick.

File: django_ai_assistant/langchain/chat_message_histories.py (hunk 2 above)
Comment (filipeximenes) [reply django_ai_assistant/langchain/chat_message_histories.py]: The langchain ones receive a uuid id that is added in the built in `add_messages` function that is called before calling this one and they are update here to the db ids a few lines bellow this.

Comment (filipeximenes) [reply django_ai_assistant/langchain/chat_message_histories.py]: I've made the change to compare ids and optimized the query to only return the ids

**Review by fjsj — APPROVED** [review]
